### ISSUE TITLE
ClickHouse Dialect does not support tuple access

### ISSUE THREAD

**Issue by mison6** [sqlfluff/sqlfluff#3816]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

ClickHouse uses `a.N` Notation, where N is an integer, to access to a tuple element.  SQLFluff misinterprets this as a Datatype followed by a numeric literal, resulting in incorrect linting errors regarding Datatypes.

### Expected Behaviour

`minmax_tuple` should not be interpreted as a Datatype.  Maybe it should be processed similarly to an array accessor?

### Observed Behaviour

For example:

```sql
WITH
(
  SELECT
    min(value) AS minimum_value,
    max(value) AS maximum_value
  FROM
    first_table
) AS minmax_tuple
SELECT
  my_group,
  count(*) AS num_rows
FROM
  other_table
WHERE
  this_value BETWEEN minmax_tuple.1 AND minmax_tuple.2
GROUP BY
  my_group
```

Parsing gives:
```
[L: 15, P:  1]      |                    whitespace:                               '  '
[L: 15, P:  3]      |                    expression:
[L: 15, P:  3]      |                        column_reference:
[L: 15, P:  3]      |                            naked_identifier:                 'this_value'
[L: 15, P: 13]      |                        whitespace:                           ' '
[L: 15, P: 14]      |                        keyword:                              'BETWEEN'
[L: 15, P: 21]      |                        whitespace:                           ' '
[L: 15, P: 22]      |                        data_type:
[L: 15, P: 22]      |                            data_type_identifier:             'minmax_tuple'
[L: 15, P: 34]      |                        numeric_literal:                      '.1'
[L: 15, P: 36]      |                        whitespace:                           ' '
[L: 15, P: 37]      |                        keyword:                              'AND'
[L: 15, P: 40]      |                        whitespace:                           ' '
[L: 15, P: 41]      |                        data_type:
[L: 15, P: 41]      |                            data_type_identifier:             'minmax_tuple'
[L: 15, P: 53]      |                        numeric_literal:                      '.2'
[L: 15, P: 55]      |                    [META] dedent:
[L: 15, P: 55]      |                newline:                                      '\n'

```

And a linting error from L063 (set in config to `pascal` to be consistent with ClickHouse datatypes)

```
L:  15 | P:  22 | L063 | Datatypes must be pascal case.
L:  15 | P:  41 | L063 | Datatypes must be pascal case.
```

### How to reproduce

```sql
-- snippet.sql
WITH
(
  SELECT
    min(value) AS minimum_value,
    max(value) AS maximum_value
  FROM
    first_table
) AS minmax_tuple
SELECT
  my_group,
  count(*) AS num_rows
FROM
  other_table
WHERE
  this_value BETWEEN minmax_tuple.1 AND minmax_tuple.2
GROUP BY
  my_group
```

 `sqlfluff parse snippet.sql --dialect clickhouse`

### Dialect

ClickHouse

### Version

1.3.0

### Configuration

```toml
[sqlfluff]
dialect = clickhouse

[sqlfluff:rules]
tab_space_size = 2

[sqlfluff:rules:L063]
# Data Types
extended_capitalisation_policy = pascal
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Ah - so this is an issue where it's parsing the `1` as a numeric literal and then saying that it's not allowed as an identifier.

To solve this we'd need to adjust the clickhouse dialect accordingly to allow it to parse properly.

So that I've got context, do you know if there are any restrictions around _when_ this format can be applied and any restrictions?

I'm thinking about edge cases for this and how we'd parse them. For example, would the following be valid:
- `my_schema.my_table.my_array_of_objs.5.my_key_within_the_obj`
- `my_array_of_arrays_of_arrays.2.3.4`

This would help to inform how we go about adding this into the dialect.

**Comment by mison6** [comment]
> do you know if there are any restrictions around when this format can be applied

I don't believe so. Any object identifier which is a tuple should be able to use that tuple accessor syntax anywhere a normal identifier could be used.

> my_schema.my_table.my_array_of_objs.5.my_key_within_the_obj

I haven't used clickhouse's [nested](https://clickhouse.com/docs/en/sql-reference/data-types/nested-data-structures/nested/) datatypes much, but a quick test makes it look like this syntax would not be valid.

```sql
CREATE TABLE my_tmp (Goals Nested(ID UInt32)) ENGINE = Memory;
INSERT INTO my_tmp Values ((1));
```

```sql
SELECT Goals.ID
FROM my_tmp

Query id: 083c5312-d654-4292-8b23-6a28715f4dd9

┌─Goals.ID─┐
│ [1]      │
└──────────┘
```

```sql
SELECT
    tuple(Goals.ID) AS test,
    test.1 AS test_1,
    test_1[1] AS test_1_1
FROM my_tmp

┌─test──┬─test_1─┬─test_1_1─┐
│ ([1]) │ [1]    │        1 │
└───────┴────────┴──────────┘
```

```sql
SELECT tuple(Goals.ID) AS test, test.1[1] FROM my_tmp

Syntax error: failed at position 39 ('['):

Expected one of: token, DoubleColon, MOD, DIV, NOT, BETWEEN, LIKE, ILIKE, NOT LIKE, NOT ILIKE, IN, NOT IN, GLOBAL IN, GLOBAL NOT IN, IS, AND, OR, QuestionMark, alias, AS, identifier, Comma, FROM, PREWHERE, WHERE, GROUP BY, WITH, HAVING, WINDOW, ORDER BY, LIMIT, OFFSET, SETTINGS, UNION, EXCEPT, INTERSECT, INTO OUTFILE, FORMAT, end of query
```

```sql
SELECT tuple(Goals) AS test, test.1.ID FROM my_tmp

Syntax error: failed at position 37 ('ID'):

Expected one of: unsigned integer, end of query
```

> my_array_of_arrays_of_arrays.2.3.4

This looks valid:

```sql
WITH ((('a', 'aa'), 'b'), 'c') AS test
SELECT
    test.1.1.2,
    (test.1).2,
    test.2

Query id: 698c0aa2-45e6-4702-af12-84f77def0dfc

┌─tupleElement(tupleElement(tupleElement(test, 1), 1), 2)─┬─tupleElement(tupleElement(test, 1), 2)─┬─tupleElement(test, 2)─┐
│ aa                                                      │ b                                      │ c                     │
└─────────────────────────────────────────────────────────┴────────────────────────────────────────┴───────────────────────┘
```

**Comment by konnectr** [comment]
@alanmcruickshank Can you find the time to implement this feature? .

**Comment by reIkaros** [comment]
I have the same problem. The simplest example is below:
`SELECT
  [1,2,3] AS arr,
  arr.1 AS dsa`

`sqlfluff parse tests.sql --dialect=clickhouse` gives:
`[L:  3, P:  3]      |                select_clause_element:
[L:  3, P:  3]      |                    expression:
[L:  3, P:  3]      |                        data_type:
[L:  3, P:  3]      |                            data_type_identifier:             'arr'
[L:  3, P:  6]      |                        numeric_literal:                      '.1'`

**Comment by reIkaros** [comment]
I have the same problem. The simplest example is below:
```sql
SELECT
  [1,2,3] AS arr,
  arr.1 AS dsa
```

`sqlfluff parse tests.sql --dialect=clickhouse` gives:
```
[L:  3, P:  3]      |                select_clause_element:
[L:  3, P:  3]      |                    expression:
[L:  3, P:  3]      |                        data_type:
[L:  3, P:  3]      |                            data_type_identifier:             'arr'
[L:  3, P:  6]      |                        numeric_literal:                      '.1'
```

And `sqlfluff fix tests.sql --dialect=clickhouse` gives:
```sql
SELECT
    [1, 2, 3] AS arr,
    arr .1 AS dsa
```
with
```
L:   3 | P:   6 | LT01 | Expected single whitespace between data type identifier
                       | and numeric literal. [layout.spacing]
```

**Comment by babaMar** [comment]
Experiencing the same problem. Where should we look to solve this exactly?
